lib/MemoryModel/MemModel.cpp

```diff
@@ -189,7 +189,7 @@ bool SymbolTableInfo::computeGepOffset(const llvm::User *V, LocationSet& ls) {
         // Handling array types, skipe array handling here
         // We treat whole array as one, then we can distinguish different field of an array of struct
         // e.g. s[1].f1 is differet from s[0].f2
-        if(isa<ArrayType>(*gi))
+        if(isa<ArrayType>(gi.getIndexedType()))
```

Comment:
Hmm...  A bit trickier than I had hoped :(

So just to make sure I understand the issue is the Operand's type?  So in your code example 
`(*gi)->dump()` is equivalent to `gi.getIndexedType()->dump()` from 3.8 to 4.0, but in the iterator the `getOperand()` in 4.0 is now pointing to the container's type rather than directly to the operand's type?

From what I saw in the code, I agree that changing the memory model would be very low in my preferences... Let me dig in a bit later tonight...  See if we can't find a better solution.